### PR TITLE
Ensure spans are truncated when encoded

### PR DESCRIPTION
**Pull request by Qard**
### What does this PR do?

This ensures trace spans are normalized and truncated properly.

### Motivation

Very large queries may be costly when encoded in the `resource.name` so we should ensure that fields are truncated if they're too long. This is also done in the agent, however the agent can still get overloaded if decoding and then re-encoding enough data with long fields.

We had already copied the truncation and normalization logic for the agentless-ci-visibility encoder, so this just reuses that logic.

### CODE REVIEW

**Comment by rochdev** [comment]
Looks like this is failing benchmarks. But more importantly, would it be a breaking change in any scenario, or is this just replicating what the agent already does? For CI the requirements may have been different.

**Comment by Qard** [comment]
Just replicating what the agent already does, so shouldn't break anything.

As for the benchmarks, it seems to have to get terminated for running too long. Is there a way to tell Circle not to do that or up the timeout somehow?

**Comment by rochdev** [comment]
@Qard You can update the benchmark to do less iterations since the timeout is already really high at 10 minutes. Not sure why they would fail for this PR specifically though.

**Comment by Qard** [comment]
Probably just close enough to the limit to go over. I just added the span stats stuff to it recently, so probably just put it into flaky territory.

**Review by rochdev — DISMISSED** [review]

File: packages/dd-trace/src/encode/0.4.js [hunk 1]
```diff
@@ -66,7 +71,8 @@ class AgentEncoder {
   _encode (bytes, trace) {
     this._encodeArrayPrefix(bytes, trace)
 
-    for (const span of trace) {
+    const events = trace.map(formatSpan)
+    for (const span of events) {
```
Comment (simon-id) [inline packages/dd-trace/src/encode/0.4.js:75]: A `.map()` just before a for loop ?  I would have just emulated the map inside the for loop, so we don't iterate on it twice:
```suggestion
    for (let span of trace) {
      span = formatSpan(span)
```

File: packages/dd-trace/src/encode/0.5.js [hunk 2]
```diff
@@ -27,7 +32,8 @@ class AgentEncoder extends BaseEncoder {
   _encode (bytes, trace) {
     this._encodeArrayPrefix(bytes, trace)
 
-    for (const span of trace) {
+    const events = trace.map(formatSpan)
+    for (const span of events) {
```
Comment (simon-id) [inline packages/dd-trace/src/encode/0.5.js:36]: Same as above comment

File: packages/dd-trace/test/encode/0.5.spec.js [hunk 3]
```diff
@@ -4,6 +4,12 @@ const msgpack = require('msgpack-lite')
 const codec = msgpack.createCodec({ int64: true })
 const id = require('../../src/id')
 
+function randString (length) {
+  return Array.from({ length }, () => {
```
Comment (simon-id) [inline packages/dd-trace/test/encode/0.5.spec.js:8]: What's the reason behind using `Array.from({ length })` instead of `new Array(length)` I wonder ?

**Review by simon-id — CHANGES_REQUESTED** [review]

**Review by simon-id — DISMISSED** [review]
(Approving because I don't want to block the PR)

File: packages/dd-trace/test/encode/0.5.spec.js (hunk 3 above)
Comment (Qard) [reply packages/dd-trace/test/encode/0.5.spec.js:8]: `Array.from({ length }, map)` is a faster way to create and map an array than separately doing `new Array(length)` and `array.map(fn)`.

**Review by rochdev — APPROVED** [review]

**Pull request merged**